File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.java
import java.io.*;
import java.util.*;

public class Main {
    static ArrayList<Integer>[] graph;
    static int[] visited;
    static int cnt = 1; 
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int r = Integer.parseInt(st.nextToken());

        graph = new ArrayList[n+1]; //그래프 초기화
        visited = new int[n+1]; //방문 배열 초기화

        for(int i = 1;i<=n;i++){
            graph[i] = new ArrayList<>();
        }

        for(int i = 0;i<m;i++){
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            graph[u].add(v);
            graph[v].add(u);
        }
        for(int i = 1;i<=n;i++){
            Collections.sort(graph[i]);
        }
        bfs(r);
        for(int i =1;i<=n;i++){
            System.out.println(visited[i]);
        }
    }
    static void bfs(int start){
       Queue<Integer> q = new LinkedList<>();
       q.add(start);
       visited[start] = cnt++;
        while(!q.isEmpty()){
            int now = q.poll();
            for(int next: graph[now]){
                if(visited[next]==0){
                    visited[next] = cnt++;
                    q.add(next);
                }
            }
        }
    }
    
}
    

File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.py
import sys
from collections import deque
input = sys.stdin.readline

def bfs(start):
    q = deque([start])
    visited[start] = 1
    cnt = 2
    while(q):
        x = q.popleft()
        for nx in graph[x]:
            if visited[nx] ==0:
                visited[nx] = cnt
                cnt +=1
                q.append(nx)

n,m,r = map(int,input().split())
graph = [[] for _ in range(n+1)]
visited = [0]*(n+1)
for _ in range(m):
    u,v = map(int,input().split())
    graph[u].append(v)
    graph[v].append(u)
for i in range(n+1):
    graph[i].sort()
bfs(r)

for i in range(1, n + 1):
    print(visited[i])